backend/src/controllers/notes.ts

```diff
@@ -140,10 +170,105 @@ export const deleteNote: RequestHandler = async (req, res, next) => {
       throw createHttpError(401, "You cannot access this note");
     }
 
-    await note.remove();
+    note.deletedAt = new Date();
+
+    await note.save();
 
     res.sendStatus(204);
   } catch (error) {
     next(error);
   }
 };
+
+export const getDeletedNotes: RequestHandler = async (req, res, next) => {
+  const authenticatedUserId = req.session.userId;
+
+  try {
+    assertIsDefined(authenticatedUserId);
+
+    const notes = await NoteModel.find({
+      userId: authenticatedUserId,
+      deletedAt: { $ne: null },
+    }).exec();
+    res.status(200).json(notes);
+  } catch (error) {
+    next(error);
+  }
+};
+
+const restoreDeletedNoteSchema = z.object({
+  noteId: z.string(),
+});
+
+export const restoreDeletedNote: RequestHandler = async (req, res, next) => {
+  const { data, error } = restoreDeletedNoteSchema.safeParse(req.params);
+  const authenticatedUserId = req.session.userId;
+
+  if (error) {
+    throw createHttpError(400, "Invalid parameters");
+  }
+
+  const { noteId } = data;
+
+  try {
+    assertIsDefined(authenticatedUserId);
+
+    if (!mongoose.isValidObjectId(noteId)) {
+      throw createHttpError(400, "Invalid note ID");
+    }
+
+    const note = await NoteModel.findById(noteId).exec();
+
+    if (!note) {
+      throw createHttpError(404, "Note not found");
+    }
+
+    if (!note.userId.equals(authenticatedUserId)) {
+      throw createHttpError(401, "You cannot access this note");
+    }
+
+    await NoteModel.findByIdAndUpdate(noteId, { $unset: { deletedAt: 1 } });
+
+    res.status(200).json(note);
+  } catch (error) {
+    next(error);
+  }
+};
+
+const permaDeleteNoteSchema = z.object({
+  noteId: z.string(),
+});
+
+export const permaDeleteNote: RequestHandler = async (req, res, next) => {
```

Comment:
Note for the future: 

Please don't use abbreviations/affixes on naming non-ambiguous things: `Delete` already means permanently erasing something from somewhere. This is not unexpected.

If that's not the case and you actually implemented a deletion functionality that's keeping a copy of the deleted thing, then you should make it obvious bc that's unexpected.